metric_learn/scml.py

```diff
@@ -0,0 +1,596 @@
+                       % type(self.n_basis))
+
+    if n_basis > uniqPairs.shape[0]:
+      n_basis = uniqPairs.shape[0]
+      warnings.warn("The selected number of basis is greater than the number "
+                    "of points, only n_basis = %d will be generated" %
+                    n_basis)
+
+    uniqPairs = X[uniqPairs]
+
+    rng = check_random_state(self.random_state)
+
+    # Select n_basis
+    selected_pairs = uniqPairs[rng.choice(uniqPairs.shape[0],
+                               size=n_basis, replace=False), :, :]
+
+    basis = selected_pairs[:, 0]-selected_pairs[:, 1]
+
+    return basis, n_basis
+
+
+class SCML_global(_BaseSCML_global, _TripletsClassifierMixin):
+  """Sparse Compositional Metric Learning (SCML)
+
+  `SCML` builds a metric as the sparse positive combination of a set of locally
+  discriminative rank-one PSD basis. This allows an optimization scheme with
+  only `K` parameters, that can be yielded with an efficient stochastic
+  composite optimization over a set of triplets constraints. Each triplet is
+  constructed as a relative distance comparison with respect to the first
+  element so that the second element is closer than the last.
+  Read more in the :ref:`User Guide <scml>`.
+  Parameters
+  ----------
+  beta: float (default=1e-5)
+      L1 regularization parameter.
+  basis : None, string or numpy array, optional (default=None)
+       Prior to set for the metric. Possible options are
+       '', and a numpy array of shape (n_basis, n_features). If
+        None an error will be raised as the basis set is esential
+        to SCML.
+       numpy array
+           A matrix of shape (n_basis, n_features), that will be used as
+           the basis set for the metric construction.
+  n_basis : int, optional
+      Number of basis to be yielded. In case it is not set it will be set based
+      on the basis numpy array. If an string option is pased to basis an error
+      wild be raised as this value will be needed.
+  gamma: float (default = 5e-3)
+      Learning rate
+  max_iter : int (default = 100000)
+      Number of iterations for the algorithm
+  output_iter : int (default = 5000)
+      Number of iterations to check current weights performance and output this
+      information in case verbose is True.
+  verbose : bool, optional
+      if True, prints information while learning
+  preprocessor : array-like, shape=(n_samples, n_features) or callable
+      The preprocessor to call to get triplets from indices. If array-like,
+      triplets will be formed like this: X[indices].
+  random_state : int or numpy.RandomState or None, optional (default=None)
+      A pseudo random number generator object or a seed for it if int.
+  Attributes
+  ----------
+  components_ : `numpy.ndarray`, shape=(n_features, n_features)
+      The linear transformation ``L`` deduced from the learned Mahalanobis
+      metric (See function `components_from_metric`.)
```

Comment:
for SCML it is `_get_components`
(maybe choose a more consistent and explicit name, like `_components_from_basis_weights`)